exploration/0474-text-vs-code.md

```diff
@@ -0,0 +1,471 @@
+to the message do not wish to reformat the message extensively.
+
+Users who have messages that include leading or trailing whitespace
+want to ensure that this whitespace is included in the translatable
+text portion of the message. Which whitespace characters are displayed at runtime
+should not be surprising.
+
+## Requirements
+
+Common things should be easy, uncommon things should be possible.
+
+Developers and translators should be able to read and write the syntax easily in a text editor.
+
+Translators (and their tools) are not software engineers, so we want our syntax
+to be as simple, robust, and non-fussy as possible.
+Multiple levels of complex nesting should be avoided,
+along with any constructs that require an excessive
+level of precision on the part of non-technical users.
+
+As MessageFormat 2 will be at best a secondary language to all its users,
+it should conform to user expectations and require as little learning as possible.
+
+The syntax should avoid footguns,
+in particular as it's passed through various tools during formatting.
+
+ASCII-compatible syntax. While support for non-ASCII characters for variable names,
+values, literals, options, and the like are important, the syntax itself should
+be restricted to ASCII characters. This allows the message to be parsed
+visually by humans even when embedded in a syntax that requires escaping.
+
+Whitespace is forgiving. We _require_ the minimum amount of whitespace and allow
+users to format or change unimportant whitespace as much as they want.
+This avoids the need for translators or tools to be super pedantic about
+formatting.
+
+## Constraints
+
+Limiting the range of characters that need to be escaped in plain text is important.
+Following past precedent,
+this design doc will only consider encapsulation styles which
+start with `{` and end with `}`.
+
+The current syntax includes some plain-ascii keywords:
+`input`, `local`, `match`, and `when`.
+
+The current syntax and active proposals include some sigil + name combinations,
+such as `:number`, `$var`, `|literal|`, `+bold`, and `@attr`.
+
+The current syntax supports unquoted literal values as operands.
+
+Messages themselves are "simple strings" and must be considered to be a single
+line of text. In many containing formats, newlines will be represented as the local
+equivalent of `\n`.
+
+## Proposed Design
+
+**Start in text mode**
+
+In this option, whitespace is trimmed from `pattern` constructs unless
+the pattern is quoted.
+
+```
+Hello world!
+
+Hello {$user}!
+
+{input $now :datetime dateStyle=long}
+Hello {$user}. Today is {$now}
+
+{local $now = {:systemGetCurrentTime :datetime dateStyle=medium}}
+Hello {$user}. Today is {$now}
+
+{match {$count :number integer=true}}
+{when 0}   Hello {$user}. Today is {$now} and you have no geese.
+{when one} Hello {$user}. Today is {$now} and you have {$count} goose.
+{when few} {  Hello {$user}, this message has spaces on the front and end.  }
```

Comment:
How would this work, parsing-wise?

```
{when few} {  Hello {$user}, this message has spaces on the front and end.  }
           ^
           how does a parser know this isn't a placeholder's open brace?
``` 

I think it may be a good idea to consider using double braces for certain fetures (e.g. for placeholders, or as pattern delimiters). Alternatively, we may want to revisit the idea of using double sigils for different meanings, e.g. `{%`, `{[`, and `{{`. See #269.

Reply:
Doubling will make the syntax harder to use/write?

In the above, the whitespace is consumed until you see the `{` (or any text). The parser knows this isn't a placeholder's opening brace only by scanning ahead. In this message, the embedded `{$user` is what resolves it. It is possible that one could reach the closing bracket in some messages and that would be the resolution.

The question here is whether we favor this syntax for its usability more than efficiency in parsing.

In my opinion, the real problem here is the `when` clause. Not only is it visually hard to distinguish but it makes `{`/`}` ambiguous (the brackets can be any of **_three_** different things). Consider instead:

```
#match {$count :number integer=true}
[when 0] This has no spaces.
[when one]      This has no spaces.
[when few] {   This has spaces quoted   }
[when *] {|  |}  This has spaces quoted on the front only as placeholder literal.
```

And it writes single line as:
```
#match {$count :number integer=true}[when 0]This has no spaces.[when one]   No sp...
...aces   [when few]{ Quoted spaces  }[when *]{|  |} Quoted spaces
```

The above still has some forward looking ambiguity:

```
{   $user :foo}   // not resolved till you see $
{   foo :foo  }   // not resolved till you see the `:`
{   foo  {$bar}}  // not resolved till you see the 2nd `{`
{   foo foo   }   // not resolved till you see the closing `}`
```

So you're right: the parser is more of an adventure once the pattern can be unquoted.

One option we discussed is that the pattern must still be quoted once we are in code mode:

```
Hello world
Hello {$user}

{input $user}
{Hello {$user}}  // because code mode
```